flash/data/data_source.py

```diff
@@ -0,0 +1,263 @@
+        return train_dataset, val_dataset, test_dataset, predict_dataset
+
+    def _generate_dataset_if_possible(
+        self,
+        running_stage: RunningStage,
+        data_pipeline: DataPipeline,
+    ) -> Optional[Union[AutoDataset, IterableAutoDataset]]:
+        data = getattr(self, f"{_STAGES_PREFIX[running_stage]}_data", None)
+        if data is not None:
+            return self.generate_dataset(data, running_stage, data_pipeline)
+
+    def generate_dataset(
+        self,
+        data,
+        running_stage: RunningStage,
+        data_pipeline: DataPipeline,
+    ) -> Optional[Union[AutoDataset, IterableAutoDataset]]:
+        mock_dataset = MockDataset()
+        with CurrentRunningStageFuncContext(running_stage, "load_data", self):
+            data = self.load_data(data, mock_dataset)  # TODO: Should actually resolve this
+
+        if has_len(data):
+            dataset = AutoDataset(data, self, running_stage, data_pipeline)
+        else:
+            dataset = IterableAutoDataset(data, self, running_stage, data_pipeline)
+        dataset.__dict__.update(mock_dataset.metadata)
+        return dataset
+
+
+class DefaultDataSource(Enum):  # TODO: This could be replaced with a data source registry that the user can add to
+
+    FOLDERS = "folders"
+    FILES = "files"
+
+    def as_type(self) -> Type[DataSource]:
+        _data_source_types = {
+            DefaultDataSource.FOLDERS: FoldersDataSource,
+            DefaultDataSource.FILES: FilesDataSource,
+        }
+        return _data_source_types[self]
+
+
+class SequenceDataSource(DataSource, ABC):
+
+    def __init__(
+        self,
+        train_inputs: Optional[Sequence[Any]] = None,
+        train_targets: Optional[Sequence[Any]] = None,
+        val_inputs: Optional[Sequence[Any]] = None,
+        val_targets: Optional[Sequence[Any]] = None,
+        test_inputs: Optional[Sequence[Any]] = None,
+        test_targets: Optional[Sequence[Any]] = None,
+        predict_inputs: Optional[Sequence[Any]] = None,
+        labels: Optional[Sequence[str]] = None
+    ):
+        super().__init__(
+            train_data=(train_inputs, train_targets),
+            val_data=(val_inputs, val_targets),
+            test_data=(test_inputs, test_targets),
+            predict_data=(predict_inputs, None),
+        )
+
+        self.labels = labels
+
+        if self.labels is not None:
+            self.preprocess.set_state(LabelsState(self.labels))
+
+    def load_data(self, data: Any, dataset: Optional[Any] = None) -> Iterable[Mapping[str, Any]]:
+        inputs, targets = data
+        if targets is None:
+            return [{'input': input} for input in inputs]
+        return [{'input': input, 'target': target} for input, target in zip(inputs, targets)]
+
+
+class FoldersDataSource(DataSource, ABC):
+
+    def __init__(
+        self,
+        train_folder: Optional[Union[str, pathlib.Path, list]] = None,
```

Comment:
I vote to only support primitive types here `list` and `list(str)`

Reply:
Yeah, typing is wrong here as it can't be a list anyway, but we should maybe just do str